packages/sync/src/modules/Term_Relationships.php

```diff
@@ -74,52 +92,55 @@ public function init_before_send() {
-	 * @see Automattic\Jetpack\Sync\Modules\Module::enqueue_all_ids_as_action
+	 * @param array  $config Full sync configuration for this sync module.
+	 * @param int    $max_items_to_enqueue Maximum number of items to enqueue.
+	 * @param object $last_object_enqueued Last object enqueued.
 	 *
-	 * @param array   $config               Full sync configuration for this sync module.
-	 * @param int     $max_items_to_enqueue Maximum number of items to enqueue.
-	 * @param boolean $state                True if full sync has finished enqueueing this module, false otherwise.
 	 * @return array Number of actions enqueued, and next module state.
+	 * @todo This method has similarities with Automattic\Jetpack\Sync\Modules\Module::enqueue_all_ids_as_action. Refactor to keep DRY.
+	 * @see Automattic\Jetpack\Sync\Modules\Module::enqueue_all_ids_as_action
 	 */
-	public function enqueue_full_sync_actions( $config, $max_items_to_enqueue, $state ) {
+	public function enqueue_full_sync_actions( $config, $max_items_to_enqueue, $last_object_enqueued ) {
 		global $wpdb;
-		$items_per_page        = 1000;
-		$chunk_count           = 0;
-		$previous_interval_end = $state ? $state : array(
-			'object_id'        => '~0',
-			'term_taxonomy_id' => '~0',
+		$term_relationships_batch_size = Settings::get_setting( 'term_relationships_batch_size' );
+		$limit                         = min( $max_items_to_enqueue * $term_relationships_batch_size, self::QUERY_LIMIT );
+		$items_enqueued_count          = 0;
+		$last_object_enqueued          = $last_object_enqueued ? $last_object_enqueued : array(
+			'object_id'        => self::MAX_INT,
+			'term_taxonomy_id' => self::MAX_INT,
 		);
-		$listener              = Listener::get_instance();
-		$action_name           = 'jetpack_full_sync_term_relationships';
 
-		// Count down from max_id to min_id so we get term relationships for the newest posts and terms first.
-		// phpcs:ignore WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
-		while ( $ids = $wpdb->get_results( "SELECT object_id, term_taxonomy_id FROM $wpdb->term_relationships WHERE object_id <= {$previous_interval_end['object_id']} AND term_taxonomy_id < {$previous_interval_end['term_taxonomy_id']} ORDER BY object_id DESC, term_taxonomy_id DESC LIMIT {$items_per_page}", ARRAY_A ) ) {
+		while ( $limit > 0 ) {
+			// Count down from max_id to min_id so we get term relationships for the newest posts and terms first.
+			$objects = $wpdb->get_results( $wpdb->prepare( "SELECT object_id, term_taxonomy_id FROM $wpdb->term_relationships WHERE ( object_id = %d AND term_taxonomy_id < %d ) OR ( object_id < %d ) ORDER BY object_id DESC, term_taxonomy_id DESC LIMIT %d", $last_object_enqueued['object_id'], $last_object_enqueued['term_taxonomy_id'], $last_object_enqueued['object_id'], $limit ), ARRAY_A );
```

Comment:
Lets write this nicer, its hard to read.

Reply:
what do you propose?

Reply:
I added a comment with a query example